submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,244 @@
+
+``` go
+type SubmarinerGwRoute struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    SubmarinerRoutePolicySpec SubmarinerRoutePolicySpec `json:"submarinerRoutePolicySpec"`
+}
+
+type SubmarinerRoutePolicySpec struct {
+    //Specifies the list of next hops to reach the remote CIDRs
+    NextHops []string `json:"nextHops"`
+
+    //Specifies the remote CIDRs available via the next hop
+    RemoteCidr []string `json:"remoteCidr"`
+}
+```
+
+#### SubmarinerNonGWRoute
+
+This CR will be created when a remote endpoint is created and there will be one created per endpoint.
+
+* NextHops - Specifies the list of next hops. In this case ,we will have only one, and it will be the transit switch IP of the zone
+where g/w node is present.
+* RemoteCIDR - Specifies the list of remote CIDRs reachable via this gateway.
+
+* In non-g/w node - If the route-agent pod is not in the same zone as Gateway node zone, send the traffic to the g/w node zone.
+* In g/w node - add route to send the traffic from other zones, destined to remote cluster to the submariner tunnel
+
+``` go
+type SubmarinerNonGWRoute struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    SubmarinerRoutePolicySpec SubmarinerRoutePolicySpec `json:"submarinerRoutePolicySpec"`
+}
+
+type SubmarinerRoutePolicySpec struct {
+    //Specifies the list of next hops to reach the remote CIDRs
+    NextHops []string `json:"nextHops"`
+
+    //Specifies the remote CIDRs available via the next hop
+    RemoteCidr []string `json:"remoteCidr"`
+}
+```
+
+## Design Details
+
+In this proposal, the plan is to remove the current `submariner-router` and switches that were added to the OVN db. The network-plugin-syncer
+shall be removed  and can be replaced by controllers in and Submariner RouteAgent. Since we have multiple ovn db to
+program, we need multiple connections. So it makes it easier to program the OVN datapath from RouteAgent, than using
+a separate pod. This new approach will work for both IC enabled and existing deployments. With this change we are planning to use the
+ovn-k8s-mp0 interface to reach the host networking stack and then the Submariner tunnel. This interface is used by OVN for host-networking
+traffic with in a cluster and will be present in every node.
+
+![UpdatedTopology](./images/submariner-ovn-ic.png)
+
+### SubmarinerRouteAgentPod
+
+The Submariner Route-agent pod running on the active gateway node will be responsible for creating the SubmarinerGWRoute CRs. It will be used
+only for OVN CNI. For every RemoteEndpointCreated event a SubmarinerGWRoute CR will be created. The nextHop will be the interface IP through
+which we can reach the cable driver. In the case of OVN it will be the IP of ovn-k8s-mp0 interface.
+
+The SubmarinerNonGWRoute CRD will also be created by Submariner Route-agent. It will be created per endpoint and will have remoterCIDRS from
```

Comment:
```suggestion
The SubmarinerNonGWRoute CRD will also be created by Submariner Route-agent running on the active gateway node. It will be created per endpoint and will have remoterCIDRS from
```

Reply:
Done.